docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,301 @@
+
+### User Stories
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+#### Story 3
+
+As a Cluster API user, I would like to have access to the scale from zero feature but my infrastructure provider
+has not yet implemented the status field updates. By using annotations on my MachineSets or MachineDeployments,
+I can utilize this feature until my infrastructure provider has completed updating their Cluster API implementation.
+
+### Implementation Details/Notes/Constraints
+
+There are 2 methods described for informing the cluster autoscaler about the resource needs of the
+nodes in each node group: through a status field on Infrastructure Machine Templates, and through
+annotations on MachineSets or MachineDeployments. The first method requires updates to a infrastructure provider's
+controllers and will require more coordination between developers and users. The second method
+requires less direct intervention from infrastructure providers and puts more resposibility on users, for
+this additional responsibility the users gain immediate access to the feature. These methods are
+mutually exclusive, and the annotations will take preference when specified.
+
+#### Infrastructure Machine Template Status Updates
+
+Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
+This field will contain the CPU, memory, and GPU resources associated with the machine described by
+the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
```

Comment:
Could `storage` also be relevant in the context of the autoscaler? (just saw that there's `corev1.ResourceName` for it)

Reply:
storage is not relevant in this context. when nodes are created then the autoscaler has a different method for detecting attached storage and storage requests.